File: Algorithms/C++/sorting/radix_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void radix_sort(vector<int> &arr)
{
    int n = (int)arr.size();
    int max_val = *max_element(arr.begin(), arr.end());

    for (int exp = 1; max_val / exp > 0; exp *= 10)
    {
        vector<int> output(n);
        vector<int> count(10);

        for (int i = 0; i < n; i++)
        {
            count[(arr[i] / exp) % 10]++;
        }

        for (int i = 1; i < 10; i++)
        {
            count[i] += count[i - 1];
        }

        for (int i = n - 1; i >= 0; i--)
        {
            output[count[(arr[i] / exp) % 10] - 1] = arr[i];
            count[(arr[i] / exp) % 10]--;
        }

        for (int i = 0; i < n; i++)
        {
            arr[i] = output[i];
        }
    }
}

int main()
{
    vector<int> array = {64, 25, 12, 22, 11};

    cout << "Original array: ";
    for (const int &num : array)
    {
        cout << num << " ";
    }
    cout << endl;

    radix_sort(array);

    cout << "Sorted array: ";
    for (const int &num : array)
    {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: Algorithms/Python/sorting/radix_sort.py
from __future__ import annotations


def radix_sort(arr: list[int]) -> None:
    if not arr:
        return

    base = 10
    max_num = max(arr)
    exp = 1

    while max_num // exp > 0:
        counting_sort(arr, exp, base)
        exp *= base


def counting_sort(arr: list[int], exp: int, base: int) -> None:
    n = len(arr)
    output = [0] * n
    count = [0] * base

    for num in arr:
        count[(num // exp) % base] += 1

    for i in range(1, base):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        output[count[(arr[i] // exp) % base] - 1] = arr[i]
        count[(arr[i] // exp) % base] -= 1

    for i in range(n):
        arr[i] = output[i]


if __name__ == "__main__":
    arr = [170, 45, 75, 90, 802, 24, 2, 66]
    radix_sort(arr)
    print(arr)  # Output: [2, 24, 45, 66, 75, 90, 170, 802]
